api/src/main/java/net/kyori/adventure/text/Component.java

```diff
@@ -1656,6 +1657,42 @@ default boolean hasStyling() {
   @Contract(pure = true)
   @NonNull Component replaceText(final @NonNull TextReplacementConfig config);
 
+  /**
+   * Calls the provided visitor for this component and all of the component's children.
+   * There is no guarantee of the order in which each component will be visited.
+   *
+   * @param visitor the component visitor
+   * @since 4.8.0
+   */
+  default void visit(final @NonNull Consumer<Component> visitor) {
+    this.visitWhile(component -> {
+      visitor.accept(component);
+      return true;
+    });
+  }
+
+  /**
+   * Calls the provided visitor for each component and all of the component's children in
+   * this component until the visitor returns {@code true}. There is no guarantee of the
+   * order in which each component will be visited.
+   *
+   * @param visitor the component visitor
+   * @since 4.8.0
+   */
+  default void visitUntil(final @NonNull Predicate<Component> visitor) {
+    this.visitWhile(visitor.negate());
+  }
+
+  /**
+   * Calls the provided visitor for each component and all of the component's children in
+   * this component while the visitor returns {@code true}. There is no guarantee of the
+   * order in which each component will be visited.
+   *
+   * @param visitor the component visitor
+   * @since 4.8.0
+   */
+  void visitWhile(final @NonNull Predicate<Component> visitor);
```

Comment:
Why don't we use the `Iterator`/`Iterable` pattern here? That would be a more standard way of traversing the tree.

```java
Iterator<Component> iterator();
Iterator<Component> iterator(/* some config? */);
```

If you want to bail early, you just stop iterating. And this would work nicely with loops.

Reply:
Nice idea - it would be relatively simple to mock up a `ComponentIterator` that nicely traverses a component structure.